examples/gossipsub-chat.rs

```diff
@@ -0,0 +1,139 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! A basic chat application with logs demonstrating libp2p and the gossipsub protocol.
+//!
+//! Using two terminal windows, start two instances. Type a message in either terminal and hit return: the
+//! message is sent and printed in the other terminal. Close with Ctrl-c.
+//!
+//! You can of course open more terminal windows and add more participants.
+//! Dialing any of the other peers will propagate the new participant to all
+//! chat members and everyone will receive all messages.
+//!
+//! In order to get the nodes to connect, take note of the listening address of the first
+//! instance and start the second with this address as the first argument. In the first terminal
+//! window, run:
+//!
+//! ```sh
+//! cargo run --example chat
+//! ```
+//!
+//! It will print the PeerId and the listening address, e.g. `Listening on
+//! "/ip4/0.0.0.0/tcp/24915"`
+//!
+//! In the second terminal window, start a new instance of the example with:
+//!
+//! ```sh
+//! cargo run --example chat -- /ip4/127.0.0.1/tcp/24915
+//! ```
+//!
+//! The two nodes should then connect.
+
+extern crate env_logger;
+extern crate futures;
+extern crate libp2p;
+extern crate tokio;
+
+use env_logger::{Builder, Env};
+use futures::prelude::*;
+use libp2p::gossipsub::{GossipsubEvent, Topic};
+use libp2p::{
+    gossipsub, identity,
+    tokio_codec::{FramedRead, LinesCodec},
+    PeerId,
+};
+use std::time::Duration;
+
+fn main() {
+    Builder::from_env(Env::default().default_filter_or("debug")).init();
+
+    // Create a random PeerId
+    let local_key = identity::Keypair::generate_ed25519();
+    let local_peer_id = PeerId::from(local_key.public());
+    println!("Local peer id: {:?}", local_peer_id);
+
+    // Set up an encrypted TCP Transport over the Mplex and Yamux protocols
+    let transport = libp2p::build_development_transport(local_key);
+
+    // Create a Floodsub/Gossipsub topic
+    let topic = Topic::new("test-net".into());
+
+    // Create a Swarm to manage peers and events
+    let mut swarm = {
+        // to set default parameters for gossipsub use:
+        // let gossipsub_config = gossipsub::GossipsubConfig::default();
+
+        // set custom gossipsub
+        let gossipsub_config = gossipsub::GossipsubConfigBuilder::new()
+            .heartbeat_interval(Duration::from_secs(10))
+            .build();
```

Comment:
This is the default, so it is already set. Multiple communicating chat instances should work. This setting would also prevent single chat services from working if the service didn't implement it's own validation. 

Are you suggesting adding the default setting just for clarity so new users are aware this can be unset?

Reply:
> Are you suggesting adding the default setting just for clarity so new users are aware this can be unset?

No, with the other changes setting `manual_propagation` to `false` by default, one can ignore my suggestion.